// @lfg/mini-engine/dist/lfg-mini-engine.js
let V = null, re = null, Ie = 4294967295, Re = 4294967295, ve = null;
function sn(e, n) {
  const t = e.replace(/\/+$/, "");
  return t.endsWith("/runtime") ? t : `${t}/${n}`;
}
function ft(e = "v1") {
  const n = typeof window < "u" && window.__LFG_ENGINE_BASE__ || "/assets/engine";
  return sn(n, e);
}
function ke(e, n) {
  V = e, re = n, console.log(
    `bound WASM module with ${Object.keys(n.Components).length} components`
  );
}
function pt() {
  return V !== null && re !== null;
}
function Pe() {
  return V;
}
function v() {
  return (V == null ? void 0 : V.Mini) ?? null;
}
function Te() {
  return re;
}
function G() {
  return Ie;
}
function Le(e) {
  Ie = e;
}
function yt() {
  return Re;
}
function mt(e) {
  Re = e;
}
function rn() {
  return ve;
}
function ht(e) {
  ve = e;
}
const an = {
  bool: { size: 1, alignment: 1 },
  u8: { size: 1, alignment: 1 },
  i8: { size: 1, alignment: 1 },
  u16: { size: 2, alignment: 2 },
  i16: { size: 2, alignment: 2 },
  u32: { size: 4, alignment: 4 },
  text_id: { size: 4, alignment: 4 },
  i32: { size: 4, alignment: 4 },
  f32: { size: 4, alignment: 4 },
  f64: { size: 8, alignment: 8 },
  u8x2: { size: 2, alignment: 1 },
  u8x3: { size: 3, alignment: 1 },
  u8x4: { size: 4, alignment: 1 },
  i8x2: { size: 2, alignment: 1 },
  i8x3: { size: 3, alignment: 1 },
  i8x4: { size: 4, alignment: 1 },
  u16x2: { size: 4, alignment: 2 },
  u16x3: { size: 6, alignment: 2 },
  u16x4: { size: 8, alignment: 2 },
  i16x2: { size: 4, alignment: 2 },
  i16x3: { size: 6, alignment: 2 },
  i16x4: { size: 8, alignment: 2 },
  u32x2: { size: 8, alignment: 4 },
  u32x3: { size: 12, alignment: 4 },
  u32x4: { size: 16, alignment: 4 },
  i32x2: { size: 8, alignment: 4 },
  i32x3: { size: 12, alignment: 4 },
  i32x4: { size: 16, alignment: 4 },
  f32x2: { size: 8, alignment: 4 },
  f32x3: { size: 12, alignment: 4 },
  f32x4: { size: 16, alignment: 4 },
  f64x2: { size: 16, alignment: 8 },
  f64x3: { size: 24, alignment: 8 },
  f64x4: { size: 32, alignment: 8 }
};
let N = [];
const Be = "__lfg_gameplay_registered_signature";
function cn() {
  return typeof window < "u" && window[Be] || "";
}
function ae(e) {
  typeof window < "u" && (window[Be] = e);
}
function he(e, n) {
  if (n <= 1) return e;
  const t = e % n;
  return t === 0 ? e : e + (n - t);
}
function Ne(e) {
  return JSON.stringify(
    e.map((n) => ({
      name: n.name,
      id: n.id,
      byteSize: n.byteSize,
      alignment: n.alignment,
      fields: Object.values(n.fields).map((t) => ({
        name: t.name,
        type: t.type,
        offset: t.offset
      }))
    }))
  );
}
function ie(e) {
  const n = e.fields ?? {}, t = {};
  let o = 0, r = 1;
  for (const [c, i] of Object.entries(n)) {
    const u = an[i.type];
    if (!u)
      throw new Error(`Unsupported gameplay component field type: ${i.type}`);
    o = he(o, u.alignment), t[c] = {
      ...i,
      name: c,
      offset: o
    }, o += u.size, r = Math.max(r, u.alignment);
  }
  const s = he(o, r);
  return {
    name: e.name,
    id: e.id >>> 0,
    typeId: e.id >>> 0,
    size: s,
    byteSize: s,
    alignment: r,
    description: e.description,
    fields: t,
    read(c, i) {
      const u = {};
      for (const p of Object.values(t))
        u[p.name] = ln(c, i + p.offset, p.type);
      return u;
    },
    write(c, i, u) {
      if (u != null)
        for (const p of Object.values(t))
          p.name in u && dn(c, i + p.offset, p.type, u[p.name]);
    }
  };
}
function xt(e) {
  return ie(e);
}
function wt(e) {
  const n = e.map((t) => ie(t));
  return {
    components: n,
    Components: n.reduce((t, o) => (t[o.name] = o, t), {}),
    ComponentIds: n.reduce((t, o) => (t[o.name] = o.id, t), {}),
    ComponentById: n.reduce((t, o) => (t[o.id] = o, t), {})
  };
}
function un() {
  return N.map((e) => ({
    ...e,
    fields: { ...e.fields }
  }));
}
function zt(e) {
  return N.find((n) => n.name === e) ?? null;
}
function gt(e) {
  return N.find((n) => n.id === e >>> 0) ?? null;
}
async function be(e = v()) {
  var o;
  if (!((o = e == null ? void 0 : e.game) != null && o.registerComponents))
    return { ok: !1, error: "Mini.game.registerComponents is not available" };
  const n = Ne(N), t = await e.game.registerComponents(N);
  return (t == null ? void 0 : t.ok) !== !1 && ae(n), t;
}
async function _t(e) {
  var t;
  N = e.map((o) => ie(o)), typeof window < "u" && (window.__lfg_gameplay_components = un());
  const n = v();
  return (t = n == null ? void 0 : n.game) != null && t.registerComponents ? Fe() ? await be(n) : { ok: !0, component_count: N.length } : (ae(""), { ok: !0, deferred: !0, component_count: N.length });
}
function Mt() {
  N = [], ae(""), typeof window < "u" && (window.__lfg_gameplay_components = []);
}
function Fe() {
  return Ne(N) !== cn();
}
function ce(e) {
  const n = e.match(/x(\d+)$/);
  return n ? Number.parseInt(n[1], 10) : 0;
}
function Se(e) {
  return ce(e) > 0 ? e.slice(0, e.lastIndexOf("x")) : e;
}
function Ue(e) {
  switch (e) {
    case "bool":
    case "u8":
    case "i8":
      return 1;
    case "u16":
    case "i16":
      return 2;
    case "f64":
      return 8;
    default:
      return 4;
  }
}
function xe(e, n, t) {
  switch (t) {
    case "bool":
      return e.getUint8(n) !== 0;
    case "u8":
      return e.getUint8(n);
    case "i8":
      return e.getInt8(n);
    case "u16":
      return e.getUint16(n, !0);
    case "i16":
      return e.getInt16(n, !0);
    case "u32":
    case "text_id":
      return e.getUint32(n, !0);
    case "i32":
      return e.getInt32(n, !0);
    case "f32":
      return e.getFloat32(n, !0);
    case "f64":
      return e.getFloat64(n, !0);
    default:
      return;
  }
}
function we(e, n, t, o) {
  switch (t) {
    case "bool":
      e.setUint8(n, o ? 1 : 0);
      return;
    case "u8":
      e.setUint8(n, Number(o) >>> 0);
      return;
    case "i8":
      e.setInt8(n, Number(o) | 0);
      return;
    case "u16":
      e.setUint16(n, Number(o) >>> 0, !0);
      return;
    case "i16":
      e.setInt16(n, Number(o) | 0, !0);
      return;
    case "u32":
    case "text_id":
      e.setUint32(n, Number(o) >>> 0, !0);
      return;
    case "i32":
      e.setInt32(n, Number(o) | 0, !0);
      return;
    case "f32":
      e.setFloat32(n, Number(o) || 0, !0);
      return;
    case "f64":
      e.setFloat64(n, Number(o) || 0, !0);
      return;
    default:
      return;
  }
}
function ln(e, n, t) {
  const o = ce(t);
  if (o <= 0)
    return xe(e, n, t);
  const r = Se(t), s = Ue(r);
  return ["x", "y", "z", "w"].slice(0, o).reduce((c, i, u) => (c[i] = xe(e, n + u * s, r), c), {});
}
function dn(e, n, t, o) {
  const r = ce(t);
  if (r <= 0) {
    we(e, n, t, o);
    return;
  }
  const s = Se(t), a = Ue(s), c = ["x", "y", "z", "w"].slice(0, r);
  !o || typeof o != "object" || c.forEach((i, u) => {
    we(
      e,
      n + u * a,
      s,
      o[i]
    );
  });
}
let b = null;
async function ue() {
  if (b) return b;
  if (window.__LFG_ASSETS__)
    return b = window.__LFG_ASSETS__, b;
  const e = window.__LFG_PROJECT_ID__;
  if (!e)
    return b = [], b;
  const n = window.__LFG_API_BASE__ || "/api", t = {};
  window.__LFG_TOKEN__ && (t.Authorization = `Bearer ${window.__LFG_TOKEN__}`);
  const o = window.__LFG_TOKEN__ ? `${n}/projects/${e}/assets` : `${n}/dev-server/${e}/assets`, r = await fetch(o, { headers: t });
  if (!r.ok)
    throw new Error(`LFG: failed to fetch assets (${r.status})`);
  return b = (await r.json()).map((s) => {
    var a;
    return {
      id: s.asset.id,
      name: s.asset.name,
      alias: s.alias,
      type: s.asset.type,
      fileUrl: s.asset.fileUrl,
      url: s.asset.fileUrl,
      thumbnailUrl: s.asset.thumbnailUrl,
      status: s.asset.status === "COMPLETED" ? "Ready" : ((a = s.asset.status) == null ? void 0 : a.toLowerCase()) || "Ready"
    };
  }), b;
}
function fn() {
  b = null;
}
const pn = 6, yn = 1, mn = "Player", te = "local";
function hn(e) {
  return `${e}_${Math.random().toString(36).slice(2, 10)}`;
}
function xn() {
  return Math.random().toString(36).slice(2, 2 + pn).toUpperCase();
}
function O(e) {
  return {
    ...e,
    players: new Map(e.players)
  };
}
function wn(e) {
  return {
    id: e.sessionId,
    name: e.playerName,
    data: null
  };
}
function ee(e) {
  return {
    sessionId: (e == null ? void 0 : e.sessionId) ?? null,
    roomCode: (e == null ? void 0 : e.roomCode) ?? null,
    phase: (e == null ? void 0 : e.phase) ?? te,
    players: (e == null ? void 0 : e.players) ?? /* @__PURE__ */ new Map(),
    world: (e == null ? void 0 : e.world) ?? null,
    mode: "local",
    authoritative: !0,
    connected: (e == null ? void 0 : e.connected) ?? !1,
    connectionState: (e == null ? void 0 : e.connectionState) ?? "idle"
  };
}
function zn(e = {}) {
  const n = /* @__PURE__ */ new Set(), t = e.createPlayerState ?? ((a) => wn(a));
  let o = ee({
    phase: e.initialPhase ?? te
  });
  function r(a) {
    for (const c of n)
      c(a);
  }
  function s(a) {
    return o = O(a), O(o);
  }
  return {
    mode: "local",
    async connect(a = {}) {
      const c = hn("local"), i = a.roomCode || xn(), u = a.playerName || mn, p = a.maxPlayers ?? yn, d = /* @__PURE__ */ new Map(), f = t({
        sessionId: c,
        playerName: u,
        roomCode: i
      });
      d.set(c, f);
      const l = s(
        ee({
          sessionId: c,
          roomCode: i,
          phase: e.initialPhase ?? te,
          players: d,
          world: e.createWorldState ? e.createWorldState({ roomCode: i, maxPlayers: p }) : null,
          connected: !0,
          connectionState: "connected"
        })
      );
      return r({ type: "connected", snapshot: l }), r({ type: "snapshot", snapshot: l }), l;
    },
    async disconnect() {
      const a = s(
        ee({
          phase: o.phase,
          world: o.world,
          connectionState: "disconnected"
        })
      );
      r({ type: "disconnected", snapshot: a });
    },
    getSnapshot() {
      return O(o);
    },
    sendCommand(a) {
      if (!o.connected || !o.sessionId || !o.roomCode)
        throw new Error(
          "Multiplayer transport is not connected; call connect() before sendCommand()."
        );
      const c = e.applyCommand ? e.applyCommand(O(o), a, {
        sessionId: o.sessionId,
        roomCode: o.roomCode
      }) : O(o), i = s({
        ...c,
        mode: "local",
        authoritative: !0,
        connected: !0,
        connectionState: "connected"
      });
      r({
        type: "command",
        command: a,
        senderId: o.sessionId,
        snapshot: i
      }), r({ type: "snapshot", snapshot: i });
    },
    subscribe(a) {
      return n.add(a), () => {
        n.delete(a);
      };
    }
  };
}
function gn() {
  if (typeof window < "u") {
    const e = window.__LFG_TRANSPORT_MODE__;
    if (e === "spacetimedb" || e === "remote") return "remote";
    if (e === "local") return "local";
    try {
      const n = new URLSearchParams(window.location.search).get("transport") || "";
      if (n === "spacetimedb" || n === "online" || n === "multiplayer" || n === "remote") return "remote";
      if (n === "local" || n === "offline") return "local";
    } catch {
    }
  }
  return "local";
}
const Et = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  createLocalTransport: zn,
  getTransportMode: gn
}, Symbol.toStringTag, { value: "Module" })), $ = /* @__PURE__ */ new Map();
function _n() {
  const e = {};
  for (const [n, t] of $)
    e[n] = t;
  return e;
}
function Mn() {
  var o;
  const e = v();
  if ((o = e == null ? void 0 : e.assets) != null && o.getLoadedAssets)
    return e.assets.getLoadedAssets(G());
  const n = [];
  let t = 0;
  for (const [r, s] of $)
    n.push({
      asset_index: t++,
      name: r,
      filename: r + ".glb",
      description: "",
      bounds_local: null,
      pivot_local: null,
      placement_point_local: null
    });
  return { count: n.length, assets: n };
}
function Ge(e, n) {
  var s;
  const t = v();
  if (!t) return { ok: !1, error: "engine not ready" };
  if ((s = t.assets) != null && s.instantiate)
    return t.assets.instantiate(G(), e, n);
  let o;
  if (typeof e == "string")
    o = $.get(e);
  else {
    let a = 0;
    for (const [, c] of $) {
      if (a === e) {
        o = c;
        break;
      }
      a++;
    }
  }
  if (o === void 0)
    return { ok: !1, error: `Asset not found: ${e}` };
  const r = t.scenes.instantiate(G(), o, {
    position: (n == null ? void 0 : n.position) || [0, 0, 0],
    rotation: (n == null ? void 0 : n.rotation) || [0, 0, 0, 1],
    scale: (n == null ? void 0 : n.scale) || [1, 1, 1]
  });
  return r != null && r.ok && t.scenes.rebuildRendererResources(G()), r;
}
async function J(e, n, t) {
  var a;
  const o = v();
  if (!(o != null && o.scenes)) return { ok: !1, error: "engine not ready" };
  const r = n || ((a = e.split("/").pop()) == null ? void 0 : a.replace(".glb", "")) || "asset", s = o.scenes.create();
  if (s === o.NULL_SCENE_HANDLE)
    return { ok: !1, error: "failed to create stub scene" };
  try {
    let c = !1, i = "";
    if (o.scenes.loadGlbFromBuffer) {
      const u = await fetch(e);
      if (!u.ok)
        return o.scenes.destroy(s), { ok: !1, error: `fetch failed: ${u.status}` };
      const p = new Uint8Array(await u.arrayBuffer()), d = o.scenes.loadGlbFromBuffer(s, p, r);
      c = typeof d == "object" ? !!d.ok : !!d, c || (i = (typeof d == "object" ? d.error : "") || "loadGlbFromBuffer failed");
    } else {
      const u = await o.scenes.loadGlbFromUrl(s, e);
      c = !!u && (typeof u == "object" ? !!u.ok : !!u), c || (i = (typeof u == "object" ? u == null ? void 0 : u.error : "") || "loadGlbFromUrl failed");
    }
    return c ? ($.set(r, s), { ok: !0, name: r, stubScene: s }) : (o.scenes.destroy(s), { ok: !1, error: i || "failed to load GLB" });
  } catch (c) {
    return o.scenes.destroy(s), { ok: !1, error: c instanceof Error ? c.message : String(c) };
  }
}
function En(e, n, t, o) {
  const r = v();
  if (!(r != null && r.scenes)) return { ok: !1, error: "engine not ready" };
  const s = t || n.replace(".glb", "") || "asset", a = r.scenes.create();
  return a === r.NULL_SCENE_HANDLE ? { ok: !1, error: "failed to create stub scene" } : r.scenes.loadGlbFromBuffer(a, e, n) ? ($.set(s, a), { ok: !0, name: s, stubScene: a }) : (r.scenes.destroy(a), { ok: !1, error: "failed to load GLB from buffer" });
}
async function je(e = {}) {
  const n = v();
  if (!(n != null && n.scenes))
    return console.warn("engine not ready; cannot load project assets"), [];
  const {
    yieldBetweenAssets: t = !0,
    yieldAfterFetch: o = !0
  } = e, r = await ue(), s = r.filter(
    (i) => i.type === "MODEL_3D" && i.status === "Ready"
  ), a = r.filter(
    (i) => i.type === "ANIMATION" && i.status === "Ready"
  );
  console.info("project assets fetched", {
    total: r.length,
    model3dReady: s.length,
    animReady: a.length
  }), o && await ne();
  const c = [];
  for (const i of s) {
    t && await ne();
    const u = i.alias || i.name;
    if ($.has(u)) {
      c.push({ asset: i, result: { ok: !0, name: u, error: "already loaded" } });
      continue;
    }
    const p = i.url || i.fileUrl;
    if (!p) {
      c.push({ asset: i, result: { ok: !1, error: "no url" } });
      continue;
    }
    const d = performance.now();
    let f;
    try {
      f = await J(p, u, i.name);
    } catch (h) {
      f = { ok: !1, error: h instanceof Error ? h.message : String(h) };
    }
    const l = Math.round(performance.now() - d);
    console.info("load finished", { assetName: u, loadDurationMs: l, ok: f.ok }), c.push({ asset: i, result: f });
  }
  for (const i of a) {
    t && await ne();
    const u = i.alias || i.name, p = i.url || i.fileUrl;
    if (p)
      try {
        const d = await J(p, u, i.name);
        c.push({ asset: i, result: d });
      } catch (d) {
        c.push({ asset: i, result: { ok: !1, error: d instanceof Error ? d.message : String(d) } });
      }
  }
  return console.info("all assets loaded", {
    total: c.length,
    ok: c.filter((i) => i.result.ok).length,
    stubScenes: $.size
  }), c;
}
const X = [];
function An() {
  var a;
  if (X.length === 0) return;
  const e = v(), n = Te(), t = Pe();
  if (!e || !n || !t) return;
  const o = (a = n.ComponentIds) == null ? void 0 : a.Transform, r = n.TransformOps;
  if (o === void 0 || !r) return;
  const s = G();
  for (const { id: c, scale: i } of X) {
    const u = e.ecs.tableGet(s, c, o);
    u && (r.setScaleX(t, u, i[0]), r.setScaleY(t, u, i[1]), r.setScaleZ(t, u, i[2]));
  }
  X.length = 0;
}
function $e(e) {
  X.length = 0;
}
async function ne() {
  if (typeof window < "u" && typeof window.requestAnimationFrame == "function") {
    await new Promise((e) => window.requestAnimationFrame(() => e()));
    return;
  }
  await new Promise((e) => setTimeout(e, 0));
}
const Cn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  getAssetScenes: _n,
  getEngineAssets: Mn,
  hideSourceAssetEntities: $e,
  instantiateAsset: Ge,
  loadAssetFromBuffer: En,
  loadAssetFromUrl: J,
  loadProjectAssetsIntoEngine: je,
  restoreSourceAssetEntities: An
}, Symbol.toStringTag, { value: "Module" }));
function In(e, n) {
  const t = v();
  if (!(t != null && t.runtime)) return !1;
  const o = typeof e == "number" ? e : G(), r = typeof e == "number" ? n : e;
  if (!r) return !1;
  try {
    return t.runtime.loadScene(o, r);
  } catch (s) {
    return console.error("failed to load scene:", s), !1;
  }
}
function Rn(e = G()) {
  const n = v();
  if (!(n != null && n.runtime)) return null;
  try {
    const t = n.runtime.saveScene(e);
    return t ? t.buffer ?? t : null;
  } catch (t) {
    return console.error("failed to save scene:", t), null;
  }
}
function vn(e = G()) {
  const n = v();
  if (!(n != null && n.scenes)) return null;
  try {
    const t = n.scenes.saveJson(e);
    if (!t) return null;
    const o = n.scenes.saveResources(e) ?? null;
    return { scene: t, resources: o };
  } catch (t) {
    return console.error("failed to save scene (full):", t), null;
  }
}
function At(e, n) {
  var o, r;
  const t = v();
  if (!(t != null && t.scenes)) return !1;
  try {
    return n.resources && !t.scenes.loadResourcesBytes(e, n.resources) && console.warn("failed to load scene resources, continuing without them"), t.scenes.loadJsonBytes(e, n.scene) ? ((r = (o = t.scenes).rebuildRendererResources) == null || r.call(o, e), t.scenes.resetRuntime(e), !0) : !1;
  } catch (s) {
    return console.error("failed to load scene (full):", s), !1;
  }
}
function Ct() {
  var n;
  const e = v();
  if (!((n = e == null ? void 0 : e.scenes) != null && n.create)) return null;
  try {
    const t = e.scenes.create();
    return t === e.NULL_SCENE_HANDLE ? null : t;
  } catch (t) {
    return console.error("failed to create scene:", t), null;
  }
}
function It(e) {
  var t;
  const n = v();
  if (!((t = n == null ? void 0 : n.scenes) != null && t.destroy)) return !1;
  try {
    return n.scenes.destroy(e);
  } catch (o) {
    return console.error("failed to destroy scene:", o), !1;
  }
}
function Rt(e) {
  var t;
  const n = v();
  if (!((t = n == null ? void 0 : n.scenes) != null && t.setMain)) return !1;
  try {
    return n.scenes.setMain(e);
  } catch (o) {
    return console.error("failed to set main scene:", o), !1;
  }
}
async function vt(e) {
  try {
    const n = await fetch(e);
    if (!n.ok || (n.headers.get("content-type") || "").includes("text/html")) return !1;
    let o = await n.arrayBuffer();
    if (o.byteLength < 8) return !1;
    const r = new Uint8Array(o)[0];
    if (r >= 65 && r <= 122)
      try {
        const s = atob(new TextDecoder().decode(o)), a = new Uint8Array(s.length);
        for (let c = 0; c < s.length; c++) a[c] = s.charCodeAt(c);
        o = a.buffer;
      } catch {
      }
    return o.byteLength < 8 ? !1 : In(o);
  } catch (n) {
    return console.error("failed to fetch scene:", n), !1;
  }
}
async function kn(e) {
  var n, t;
  switch (e.type) {
    case "scene_reset":
      return { ok: !0, handled: !1, reloadRecommended: !0, reason: "scene_reset_requires_reload" };
    case "engine_restart": {
      if (typeof window < "u") {
        const o = window.__gc_restart_mini_playback;
        if (typeof o == "function" && await o())
          return { ok: !0, handled: !0, reloadRecommended: !1 };
      }
      return { ok: !0, handled: !1, reloadRecommended: !0, reason: "engine_restart_requires_reload" };
    }
    case "save_scene": {
      const o = e.includeResources !== !1, r = ((n = rn()) == null ? void 0 : n(e.sceneKey, o)) ?? (o ? vn() : (() => {
        const c = Rn();
        return c ? { scene: new Uint8Array(c), resources: null } : null;
      })());
      if (!r) return { ok: !1, handled: !1, reason: "save_scene_failed" };
      let s = "";
      for (const c of r.scene) s += String.fromCharCode(c);
      let a = "";
      for (const c of r.resources ?? []) a += String.fromCharCode(c);
      return {
        ok: !0,
        handled: !0,
        details: {
          sceneBase64: btoa(s),
          sceneByteLength: r.scene.byteLength,
          resourcesBase64: r.resources ? btoa(a) : null,
          resourcesByteLength: ((t = r.resources) == null ? void 0 : t.byteLength) ?? 0
        }
      };
    }
    case "asset_added":
    case "asset_updated":
    case "asset_status_updated":
      return Pn(e);
    default:
      return { ok: !1, handled: !1, reloadRecommended: !0, reason: "unsupported_preview_event", details: e };
  }
}
async function Pn(e) {
  var a;
  const n = (a = e.status) == null ? void 0 : a.toLowerCase();
  if (n && n !== "ready" && n !== "completed")
    return { ok: !0, handled: !1, reloadRecommended: !1, reason: "asset_not_ready", details: { status: e.status } };
  const t = e.url || null;
  if (!t) return { ok: !0, handled: !1, reloadRecommended: !0, reason: "missing_asset_url", details: e };
  const o = e.alias || e.name || "asset", r = await J(t, o, e.description || e.name || o);
  if (!r.ok || typeof r.asset_index != "number")
    return { ok: !1, handled: !1, reloadRecommended: !0, reason: "asset_load_failed", details: r };
  const s = Ge(r.asset_index, { position: [0, 0.6, 0], scale: [1, 1, 1] });
  return s.ok ? { ok: !0, handled: !0, reloadRecommended: !1, details: { loadResult: r, instantiateResult: s } } : { ok: !1, handled: !1, reloadRecommended: !0, reason: "asset_instantiate_failed", details: s };
}
let ze = !1;
typeof window < "u" && !ze && (ze = !0, window.addEventListener("message", (e) => {
  const n = e.data;
  !n || n.type !== "gc-preview-event" || !n.event || kn(n.event).then((t) => {
    var r;
    const o = {
      type: "gc-preview-event-result",
      requestId: n.requestId,
      ok: t.ok,
      handled: t.handled,
      reloadRecommended: t.reloadRecommended,
      reason: t.reason,
      details: t.details
    };
    (r = e.source) == null || r.postMessage(o, { targetOrigin: "*" });
  }).catch((t) => {
    var o;
    (o = e.source) == null || o.postMessage({
      type: "gc-preview-event-result",
      requestId: n.requestId,
      ok: !1,
      handled: !1,
      reloadRecommended: !0,
      reason: "preview_event_threw",
      details: t instanceof Error ? { message: t.message } : { error: String(t) }
    }, { targetOrigin: "*" });
  });
}));
const Tn = "v1", K = typeof window < "u" ? window : null;
let W = null, ge = !1;
function qe(e) {
  var n, t, o, r;
  return typeof ((t = (n = e == null ? void 0 : e.Mini) == null ? void 0 : n.scenes) == null ? void 0 : t.ensureDefaultMaterial) == "function" && typeof ((r = (o = e == null ? void 0 : e.Mini) == null ? void 0 : o.scenes) == null ? void 0 : r.addRuntimeMesh) == "function";
}
function He() {
  const n = (typeof window < "u" && window.__LFG_ENGINE_BASE__ || "/assets/engine").replace(/\/+$/, "");
  return n.endsWith("/runtime") ? n : `${n}/${Tn}`;
}
function Ln() {
  const e = (K == null ? void 0 : K.__gc_mini_runtime_context) ?? null;
  return e ? qe(e.module) ? e : (console.warn("discarding stale mini runtime context without required scene helpers"), null) : null;
}
function Bn(e) {
  K && (K.__gc_mini_runtime_context = e);
}
async function Nn() {
  return await import(
    /* @vite-ignore */
    `${He()}/components.js`
  );
}
async function bn(e) {
  const n = He(), t = window.Module;
  if (t != null && t.calledRun && qe(t))
    return t.canvas = e, t;
  t != null && t.calledRun && console.warn("existing mini module is missing required scene helpers; reloading engine scripts");
  const o = 256, r = 32768, s = new WebAssembly.Memory({
    initial: o,
    maximum: r,
    shared: !0
  });
  return window.__LFG_WASM_MEMORY__ = s, window.Module = {
    ...window.Module,
    canvas: e,
    wasmMemory: s,
    locateFile(a) {
      return `${n}/${a}`;
    }
  }, window.Module.disableMiniDevStub = !0, await Fn(`${n}/mini.js`), await new Promise((a, c) => {
    const i = window.setTimeout(() => {
      c(new Error("Mini engine runtime initialization timed out"));
    }, 15e3), u = window.Module;
    function p(l) {
      return !l.Mini && typeof l.bindMiniModule == "function" && l.bindMiniModule(l), l.canvas = e, l;
    }
    if (u != null && u.calledRun) {
      clearTimeout(i), a(p(u));
      return;
    }
    const d = u == null ? void 0 : u.onRuntimeInitialized;
    if (u) {
      u.onRuntimeInitialized = () => {
        clearTimeout(i), d == null || d(), a(p(window.Module));
      };
      return;
    }
    const f = setInterval(() => {
      var l;
      (l = window.Module) != null && l.calledRun && (clearInterval(f), clearTimeout(i), a(p(window.Module)));
    }, 100);
  });
}
function Fn(e) {
  return new Promise((n, t) => {
    const o = document.createElement("script");
    o.src = e, o.async = !0, o.onload = () => n(), o.onerror = () => t(new Error(`Failed to load ${e}`)), document.head.appendChild(o);
  });
}
function _e(e) {
  const n = document.getElementById("status");
  n && (n.textContent = e);
}
async function Sn(e) {
  return new Promise((n) => setTimeout(n, e));
}
async function Un(e, n) {
  for (let t = 1; t <= 3; t++) {
    fn();
    const o = await ue();
    console.info("hydrate attempt", t, "— project assets:", o.length);
    const r = await je(), s = r.filter(({ result: a }) => a.ok).length;
    if (console.info("loaded", s, "of", r.length, "assets"), o.length > 0 || t === 3)
      return s;
    await Sn(1e3);
  }
  return 0;
}
function Gn(e) {
  const n = (e == null ? void 0 : e.canvas) ?? document.getElementById("canvas");
  if (!(n instanceof HTMLCanvasElement))
    throw new Error("Missing #canvas element");
  return n;
}
function De(e) {
  const n = e.getBoundingClientRect(), t = window.devicePixelRatio || 1;
  return {
    width: Math.max(1, Math.round(n.width * t)),
    height: Math.max(1, Math.round(n.height * t)),
    devicePixelRatio: t
  };
}
function jn(e) {
  ge || e != null && e.editor || (ge = !0, console.warn(
    "[LFG Boot] Standalone mini-engine runtime does not install Mini.editor. Editor helpers are only available when the LFG platform or another host injects the @lfg/editor bridge."
  ));
}
async function $n(e) {
  const n = Gn(e), t = Ln();
  if (t)
    return t.canvas = n, t.module.canvas = n, ke(t.module, t.bindings), Le(t.sceneHandle), t;
  if (W) {
    const o = await W;
    return o.canvas = n, o.module.canvas = n, o;
  }
  W = qn(n);
  try {
    return await W;
  } finally {
    W = null;
  }
}
async function qn(e) {
  _e("Loading mini engine runtime...");
  const [n, t] = await Promise.all([
    Nn(),
    bn(e)
  ]);
  ke(t, n), typeof window < "u" && (window.__gc_engine_bindings = n, window.__miniComponentIds = n.ComponentIds, window.__miniTransformOps = n.TransformOps);
  const o = v();
  if (!o)
    throw new Error("Module.Mini not available — mini.js may have failed to initialize");
  if (Fe()) {
    const i = await be(o);
    (i == null ? void 0 : i.ok) === !1 && console.warn("gameplay component sync returned error (may already be registered):", i.error);
  }
  const r = De(e);
  await o.renderer.initAsync(r.width, r.height, r.devicePixelRatio);
  const s = o.scenes.create();
  if (s === o.NULL_SCENE_HANDLE)
    throw new Error("Failed to create scene");
  Le(s), o.scenes.loadResourceCache(s, "/assets/resources");
  const a = {
    canvas: e,
    bindings: n,
    module: t,
    mini: o,
    sceneHandle: s
  };
  Bn(a);
  const c = await Un();
  return $e(), o.scenes.setMain(s), _e(
    c > 0 ? `Preview ready. Loaded ${c} project asset(s).` : "Preview ready."
  ), a;
}
function Hn(e) {
  const n = De(e.canvas);
  return e.module.canvas = e.canvas, e.mini.renderer.resize(n.width, n.height, n.devicePixelRatio), n;
}
function Dn(e, n) {
  e.module.canvas = e.canvas, e.mini.renderer.frame(n);
}
let Me = !1;
function Vn() {
  if (Me) return;
  Me = !0;
  const e = window;
  e.__gc_keysDown || (e.__gc_keysDown = /* @__PURE__ */ new Set()), e.__gc_keysPressed || (e.__gc_keysPressed = /* @__PURE__ */ new Set()), window.addEventListener("keydown", (n) => {
    n.repeat || (e.__gc_keysDown.add(n.code), e.__gc_keysPressed.add(n.code));
  }), window.addEventListener("keyup", (n) => {
    e.__gc_keysDown.delete(n.code);
  });
}
async function kt(e, n) {
  const t = typeof window < "u" && window.__lfg_editor;
  if (t != null && t.registerGameplay && (t != null && t.initMiniEditorBridge)) {
    t.registerGameplay({
      begin: e.begin,
      tick: e.tick,
      end: e.end ?? (() => {
      })
    }), await t.initMiniEditorBridge();
    return;
  }
  Vn();
  const o = await $n(n), { module: r, mini: s, sceneHandle: a } = o;
  jn(s);
  const { getAssetScenes: c } = await Promise.resolve().then(() => Cn), i = c();
  await e.begin(r, s, a, i);
  const u = () => Hn(o);
  window.addEventListener("resize", u), u();
  let p = performance.now();
  function d(f) {
    const l = Math.min((f - p) / 1e3, 0.03333333333333333);
    p = f, e.tick(r, s, a, l), Dn(o, l), requestAnimationFrame(d);
  }
  requestAnimationFrame(d);
}
let M = null, P = null;
const U = /* @__PURE__ */ new Map(), T = /* @__PURE__ */ new Map();
let Ve = 1, le = !1;
function Pt() {
  le = !0;
}
function Tt() {
  le = !1;
}
function Oe() {
  return M ? (M.state === "suspended" && M.resume(), M) : (M = new AudioContext(), P = M.createGain(), P.connect(M.destination), M);
}
function Lt() {
  (M == null ? void 0 : M.state) === "suspended" && M.resume();
}
function Bt() {
  return M;
}
async function On(e, n) {
  if (M || Oe(), !M) return null;
  try {
    const t = await fetch(e);
    if (!t.ok)
      return console.warn(`failed to fetch ${e}: ${t.status}`), null;
    const o = await t.arrayBuffer();
    return Wn(o, n);
  } catch (t) {
    return console.error(`error loading audio from ${e}:`, t), null;
  }
}
async function Wn(e, n) {
  if (M || Oe(), !M) return null;
  try {
    const t = await M.decodeAudioData(e);
    return U.set(n, t), {
      name: n,
      duration: t.duration,
      sampleRate: t.sampleRate,
      channels: t.numberOfChannels
    };
  } catch (t) {
    return console.error(`error decoding audio "${n}":`, t), null;
  }
}
async function Nt() {
  const e = await ue(), n = e.filter(
    (o) => o.type === "AUDIO" && o.status === "Ready"
  );
  console.info("loading project audio assets", {
    total: e.length,
    audioReady: n.length
  });
  const t = [];
  for (const o of n) {
    const r = o.url || o.fileUrl;
    if (!r) continue;
    const s = o.alias || o.name;
    if (U.has(s)) {
      const c = U.get(s);
      t.push({
        name: s,
        duration: c.duration,
        sampleRate: c.sampleRate,
        channels: c.numberOfChannels
      });
      continue;
    }
    const a = await On(r, s);
    a && t.push(a);
  }
  return t;
}
function Yn(e, n = {}) {
  if (!le) return -1;
  if (!M || !P)
    return console.warn("audio not initialized - call initAudio() first"), -1;
  const t = U.get(e);
  if (!t)
    return console.warn(`sound "${e}" not loaded`), -1;
  M.state === "suspended" && M.resume();
  const o = M.createBufferSource();
  o.buffer = t, o.loop = n.loop ?? !1, o.playbackRate.value = n.rate ?? 1;
  const r = M.createGain();
  r.gain.value = n.volume ?? 1, o.connect(r), r.connect(P);
  const s = Ve++, a = n.offset ?? 0, c = {
    id: s,
    name: e,
    source: o,
    gain: r,
    loop: o.loop,
    startedAt: M.currentTime,
    pausedAt: 0,
    offset: a,
    rate: o.playbackRate.value,
    paused: !1
  };
  return T.set(s, c), o.onended = () => {
    var i;
    T.delete(s), (i = n.onEnd) == null || i.call(n);
  }, o.start(0, a), s;
}
let B = -1;
function bt(e, n = 1) {
  return B !== -1 && de(B), B = Yn(e, { loop: !0, volume: n }), B;
}
function Ft() {
  B !== -1 && (de(B), B = -1);
}
function de(e) {
  const n = T.get(e);
  if (n) {
    try {
      n.source.onended = null, n.source.stop();
    } catch {
    }
    T.delete(e), e === B && (B = -1);
  }
}
function Kn() {
  for (const [e] of T)
    de(e);
  B = -1;
}
function St(e) {
  const n = T.get(e);
  if (!n || n.paused || !M) return;
  const t = (M.currentTime - n.startedAt) * n.rate;
  n.pausedAt = n.offset + t, n.paused = !0;
  try {
    n.source.onended = null, n.source.stop();
  } catch {
  }
}
function Ut(e) {
  const n = T.get(e);
  if (!n || !n.paused || !M || !P) return;
  const t = U.get(n.name);
  if (!t) return;
  const o = M.createBufferSource();
  o.buffer = t, o.loop = n.loop, o.playbackRate.value = n.rate, o.connect(n.gain), o.onended = () => {
    T.delete(e);
  }, o.start(0, n.pausedAt % t.duration), n.source = o, n.startedAt = M.currentTime, n.offset = n.pausedAt, n.paused = !1;
}
function Gt(e) {
  P && (P.gain.value = Math.max(0, Math.min(1, e)));
}
function jt() {
  return (P == null ? void 0 : P.gain.value) ?? 1;
}
function $t(e, n) {
  const t = T.get(e);
  t && (t.gain.gain.value = Math.max(0, Math.min(1, n)));
}
function qt(e) {
  return U.has(e);
}
function Ht(e) {
  const n = U.get(e);
  return n ? {
    name: e,
    duration: n.duration,
    sampleRate: n.sampleRate,
    channels: n.numberOfChannels
  } : null;
}
function Dt() {
  return Array.from(U.keys());
}
function Vt(e) {
  const n = T.get(e);
  return n != null && !n.paused;
}
function Ot() {
  return T.size;
}
function Wt() {
  Kn(), U.clear(), M && (M.close(), M = null), P = null, Ve = 1;
}
const We = 1e-6;
function fe(e, n, t) {
  return Math.max(n, Math.min(t, e));
}
function F(e, n, t) {
  return e + (n - e) * t;
}
function Ye(e) {
  return e * (Math.PI / 180);
}
function Ke(e) {
  return e * (180 / Math.PI);
}
function Yt(e, n) {
  return Math.hypot(e, n);
}
function Qn(e = 0, n = 0, t = 0) {
  return { x: e, y: n, z: t };
}
function Qe(e) {
  return { x: e.x, y: e.y, z: e.z };
}
function oe(e, n) {
  return { x: e.x + n.x, y: e.y + n.y, z: e.z + n.z };
}
function j(e, n) {
  return { x: e.x - n.x, y: e.y - n.y, z: e.z - n.z };
}
function se(e, n) {
  return { x: e.x * n, y: e.y * n, z: e.z * n };
}
function Xn(e, n) {
  return e.x * n.x + e.y * n.y + e.z * n.z;
}
function D(e, n) {
  return {
    x: e.y * n.z - e.z * n.y,
    y: e.z * n.x - e.x * n.z,
    z: e.x * n.y - e.y * n.x
  };
}
function pe(e) {
  return Math.hypot(e.x, e.y, e.z);
}
function k(e, n = { x: 0, y: 1, z: 0 }) {
  const t = pe(e);
  return t <= We ? Qe(n) : { x: e.x / t, y: e.y / t, z: e.z / t };
}
function Jn(e, n, t) {
  return { x: F(e.x, n.x, t), y: F(e.y, n.y, t), z: F(e.z, n.z, t) };
}
function Zn(e, n) {
  return pe(j(e, n));
}
function Xe(e = 0, n = 0, t = 0, o = 1) {
  return { x: e, y: n, z: t, w: o };
}
function Je() {
  return Xe(0, 0, 0, 1);
}
function et(e) {
  return { x: e.x, y: e.y, z: e.z, w: e.w };
}
function S(e) {
  const n = Math.hypot(e.x, e.y, e.z, e.w);
  return n <= We ? Je() : {
    x: e.x / n,
    y: e.y / n,
    z: e.z / n,
    w: e.w / n
  };
}
function nt(e, n) {
  return {
    x: e.w * n.x + e.x * n.w + e.y * n.z - e.z * n.y,
    y: e.w * n.y - e.x * n.z + e.y * n.w + e.z * n.x,
    z: e.w * n.z + e.x * n.y - e.y * n.x + e.z * n.w,
    w: e.w * n.w - e.x * n.x - e.y * n.y - e.z * n.z
  };
}
function tt(e, n) {
  const t = k(e, { x: 0, y: 1, z: 0 }), o = n * 0.5, r = Math.sin(o);
  return S({
    x: t.x * r,
    y: t.y * r,
    z: t.z * r,
    w: Math.cos(o)
  });
}
function ot(e, n) {
  const t = e * 0.5, o = n * 0.5, r = Math.sin(t), s = Math.cos(t), a = Math.sin(o), c = Math.cos(o);
  return S({
    x: a * s,
    y: c * r,
    z: -a * r,
    w: c * s
  });
}
function Ze(e, n) {
  const t = { x: e.x, y: e.y, z: e.z }, o = D(t, n), r = D(t, o);
  return oe(n, oe(se(o, 2 * e.w), se(r, 2)));
}
function st(e, n, t) {
  let o = n.x, r = n.y, s = n.z, a = n.w, c = e.x * o + e.y * r + e.z * s + e.w * a;
  if (c < 0 && (o = -o, r = -r, s = -s, a = -a, c = -c), c > 0.9995)
    return S({
      x: F(e.x, o, t),
      y: F(e.y, r, t),
      z: F(e.z, s, t),
      w: F(e.w, a, t)
    });
  const i = Math.acos(fe(c, -1, 1)), u = i * t, p = Math.sin(u), d = Math.sin(i), f = Math.cos(u) - c * p / d, l = p / d;
  return {
    x: e.x * f + o * l,
    y: e.y * f + r * l,
    z: e.z * f + s * l,
    w: e.w * f + a * l
  };
}
function en(e, n = { x: 0, y: 1, z: 0 }) {
  const t = k(
    { x: -e.x, y: -e.y, z: -e.z },
    { x: 0, y: 0, z: 1 }
  ), o = k(D(n, t), { x: 1, y: 0, z: 0 }), r = D(t, o), s = o.x, a = r.x, c = t.x, i = o.y, u = r.y, p = t.y, d = o.z, f = r.z, l = t.z, h = s + u + l;
  if (h > 0) {
    const y = 0.5 / Math.sqrt(h + 1);
    return S({
      x: (f - p) * y,
      y: (c - d) * y,
      z: (i - a) * y,
      w: 0.25 / y
    });
  }
  if (s > u && s > l) {
    const y = 2 * Math.sqrt(1 + s - u - l);
    return S({
      x: 0.25 * y,
      y: (a + i) / y,
      z: (c + d) / y,
      w: (f - p) / y
    });
  }
  if (u > l) {
    const y = 2 * Math.sqrt(1 + u - s - l);
    return S({
      x: (a + i) / y,
      y: 0.25 * y,
      z: (p + f) / y,
      w: (c - d) / y
    });
  }
  const m = 2 * Math.sqrt(1 + l - s - u);
  return S({
    x: (c + d) / m,
    y: (p + f) / m,
    z: 0.25 * m,
    w: (i - a) / m
  });
}
function Kt(e) {
  return { x: -Math.sin(e), y: 0, z: -Math.cos(e) };
}
function Qt(e) {
  return { x: Math.cos(e), y: 0, z: -Math.sin(e) };
}
function rt(e, n) {
  return Ze(n, e);
}
const Xt = {
  create: Qn,
  clone: Qe,
  add: oe,
  sub: j,
  scale: se,
  dot: Xn,
  cross: D,
  length: pe,
  normalize: k,
  lerp: Jn,
  distance: Zn,
  applyQuaternion: rt
}, Jt = {
  create: Xe,
  identity: Je,
  clone: et,
  normalize: S,
  multiply: nt,
  fromAxisAngle: tt,
  fromYawPitch: ot,
  rotateVector: Ze,
  slerp: st,
  lookRotation: en
}, Zt = {
  clamp: fe,
  lerp: F,
  degToRad: Ye,
  radToDeg: Ke
}, eo = {
  clamp: fe,
  lerp: F,
  degToRad: Ye,
  radToDeg: Ke
}, no = {
  lookRotation: en
}, to = S, nn = 28, Y = {
  baseColor: [1, 1, 1, 1],
  roughness: 0.82,
  metallic: 0,
  normalScale: 1,
  alphaCutoff: 0
};
function tn() {
  const e = Te();
  if (!e)
    throw new Error(
      "Engine not initialised — call bootMiniEngineRuntime() or startGame() before using ECS helpers"
    );
  return e;
}
function at() {
  const e = Pe();
  if (!e)
    throw new Error("WASM module not loaded");
  return e;
}
function R(e, n, t, o, r = 4294967295) {
  return { position: e, uv: n, normal: t, tangent: o, color: r >>> 0, tangentSign: 1 };
}
function Ee(e, n = 3) {
  function t(o) {
    const s = Math.max(-1, Math.min(1, o)) * 0.5 + 0.5;
    return Math.max(0, Math.min(1023, Math.round(s * 1023)));
  }
  return (t(e[0]) | t(e[1]) << 10 | t(e[2]) << 20 | (n & 3) << 30) >>> 0;
}
function oo(e, n, t, o = 255) {
  return (e & 255 | (n & 255) << 8 | (t & 255) << 16 | (o & 255) << 24) >>> 0;
}
function Ae(e) {
  let n = e;
  return (n < -8 || n > 8) && (n = (n % 16 + 16) % 16), Math.max(0, Math.min(65535, Math.round((n + 8) / 16 * 65535)));
}
function it(e, n, t, o, r, s, a = 4294967295, c = 1) {
  const i = n * nn;
  e.setFloat32(i + 0, t.x, !0), e.setFloat32(i + 4, t.y, !0), e.setFloat32(i + 8, t.z, !0), e.setUint16(i + 12, Ae(s.x), !0), e.setUint16(i + 14, Ae(s.y), !0), e.setUint32(i + 16, Ee([o.x, o.y, o.z], 3), !0), e.setUint32(
    i + 20,
    Ee([r.x, r.y, r.z], c < 0 ? 0 : 3),
    !0
  ), e.setUint32(i + 24, a >>> 0, !0);
}
function so() {
  return { vertices: [], indices: [] };
}
function ro(e) {
  const n = new Uint8Array(e.vertices.length * nn), t = new DataView(n.buffer);
  for (let o = 0; o < e.vertices.length; ++o) {
    const r = e.vertices[o];
    it(t, o, r.position, r.normal, r.tangent, r.uv, r.color, r.tangentSign);
  }
  return {
    vertexBytes: n,
    indices: new Uint16Array(e.indices),
    vertexCount: e.vertices.length
  };
}
function ct(e) {
  return e !== null && typeof e == "object" && Number.isFinite(e.x) && Number.isFinite(e.y) && Number.isFinite(e.z);
}
function A(e, n) {
  if (!ct(e))
    throw new Error(`${n} must be a vec3-like object with finite x/y/z numbers`);
}
function ye(e = "xz") {
  switch (e) {
    case "xz":
      return {
        normal: { x: 0, y: 1, z: 0 },
        tangent: { x: 1, y: 0, z: 0 },
        bitangent: { x: 0, y: 0, z: -1 }
      };
    case "xy":
      return {
        normal: { x: 0, y: 0, z: 1 },
        tangent: { x: 1, y: 0, z: 0 },
        bitangent: { x: 0, y: 1, z: 0 }
      };
    case "yz":
      return {
        normal: { x: 1, y: 0, z: 0 },
        tangent: { x: 0, y: 0, z: 1 },
        bitangent: { x: 0, y: -1, z: 0 }
      };
    default:
      throw new Error(`unsupported planar orientation '${e}', expected 'xz', 'xy', or 'yz'`);
  }
}
function me(e, n, t, o, r) {
  return {
    x: e.x + n.x * o + t.x * r,
    y: e.y + n.y * o + t.y * r,
    z: e.z + n.z * o + t.z * r
  };
}
function ut(e) {
  return A(e.center, "appendBox() options.center"), A(e.size, "appendBox() options.size"), {
    center: e.center,
    resolvedHalfExtents: {
      x: e.size.x * 0.5,
      y: e.size.y * 0.5,
      z: e.size.z * 0.5
    },
    resolvedColor: e.color
  };
}
function H(e, { p0: n, p1: t, p2: o, p3: r, color: s, uvScale: a = 1 }) {
  A(n, "appendQuad() options.p0"), A(t, "appendQuad() options.p1"), A(o, "appendQuad() options.p2"), A(r, "appendQuad() options.p3");
  const c = k(D(j(t, n), j(o, n))), i = k(j(t, n)), u = e.vertices.length;
  e.vertices.push(R(n, { x: 0, y: 0 }, c, i, s)), e.vertices.push(R(t, { x: a, y: 0 }, c, i, s)), e.vertices.push(R(o, { x: a, y: a }, c, i, s)), e.vertices.push(R(r, { x: 0, y: a }, c, i, s)), e.indices.push(u, u + 1, u + 2, u, u + 2, u + 3);
}
function Q(e, { p0: n, p1: t, p2: o, color: r }) {
  A(n, "appendTriangle() options.p0"), A(t, "appendTriangle() options.p1"), A(o, "appendTriangle() options.p2");
  const s = k(D(j(t, n), j(o, n))), a = k(j(t, n)), c = e.vertices.length;
  e.vertices.push(R(n, { x: 0, y: 0 }, s, a, r)), e.vertices.push(R(t, { x: 1, y: 0 }, s, a, r)), e.vertices.push(R(o, { x: 0.5, y: 1 }, s, a, r)), e.indices.push(c, c + 1, c + 2);
}
function ao(e, n) {
  const { center: t, resolvedHalfExtents: o, resolvedColor: r } = ut(n), s = {
    nxnynz: { x: t.x - o.x, y: t.y - o.y, z: t.z - o.z },
    pxnynz: { x: t.x + o.x, y: t.y - o.y, z: t.z - o.z },
    pxpynz: { x: t.x + o.x, y: t.y + o.y, z: t.z - o.z },
    nxpynz: { x: t.x - o.x, y: t.y + o.y, z: t.z - o.z },
    nxnypz: { x: t.x - o.x, y: t.y - o.y, z: t.z + o.z },
    pxnypz: { x: t.x + o.x, y: t.y - o.y, z: t.z + o.z },
    pxpypz: { x: t.x + o.x, y: t.y + o.y, z: t.z + o.z },
    nxpypz: { x: t.x - o.x, y: t.y + o.y, z: t.z + o.z }
  };
  H(e, { p0: s.nxnypz, p1: s.pxnypz, p2: s.pxpypz, p3: s.nxpypz, color: r }), H(e, { p0: s.pxnynz, p1: s.nxnynz, p2: s.nxpynz, p3: s.pxpynz, color: r }), H(e, { p0: s.nxnynz, p1: s.nxnypz, p2: s.nxpypz, p3: s.nxpynz, color: r }), H(e, { p0: s.pxnypz, p1: s.pxnynz, p2: s.pxpynz, p3: s.pxpypz, color: r }), H(e, { p0: s.nxpynz, p1: s.nxpypz, p2: s.pxpypz, p3: s.pxpynz, color: r }), H(e, { p0: s.nxnynz, p1: s.pxnynz, p2: s.pxnypz, p3: s.nxnypz, color: r });
}
function io(e, { baseCenter: n, size: t, color: o }) {
  A(n, "appendPyramid() options.baseCenter"), A(t, "appendPyramid() options.size");
  const r = t.x * 0.5, s = t.z * 0.5, a = n.y, c = n.y + t.y, i = { x: n.x - r, y: a, z: n.z - s }, u = { x: n.x + r, y: a, z: n.z - s }, p = { x: n.x + r, y: a, z: n.z + s }, d = { x: n.x - r, y: a, z: n.z + s }, f = { x: n.x, y: c, z: n.z };
  Q(e, { p0: i, p1: u, p2: f, color: o }), Q(e, { p0: u, p1: p, p2: f, color: o }), Q(e, { p0: p, p1: d, p2: f, color: o }), Q(e, { p0: d, p1: i, p2: f, color: o }), H(e, { p0: i, p1: d, p2: p, p3: u, color: o });
}
function co(e, {
  center: n,
  width: t = 1,
  height: o = 1,
  widthSegments: r = 1,
  heightSegments: s = 1,
  orientation: a = "xz",
  color: c
}) {
  A(n, "appendPlane() options.center"), r = Math.max(1, Math.floor(r)), s = Math.max(1, Math.floor(s));
  const i = t * 0.5, u = o * 0.5, p = r + 1, d = s + 1, f = t / r, l = o / s, { normal: h, tangent: m, bitangent: y } = ye(a), g = [];
  for (let x = 0; x < d; ++x) {
    const _ = [], w = x * l - u;
    for (let z = 0; z < p; ++z) {
      const E = z * f - i, C = me(n, m, y, E, -w);
      _.push(e.vertices.length), e.vertices.push(
        R(C, { x: z / r, y: 1 - x / s }, h, m, c)
      );
    }
    g.push(_);
  }
  for (let x = 0; x < s; ++x)
    for (let _ = 0; _ < r; ++_) {
      const w = g[x][_], z = g[x + 1][_], E = g[x + 1][_ + 1], C = g[x][_ + 1];
      e.indices.push(w, z, C, z, E, C);
    }
}
function lt(e, {
  center: n,
  radius: t = 1,
  segments: o = 32,
  thetaStart: r = 0,
  thetaLength: s = Math.PI * 2,
  orientation: a = "xz",
  color: c
}) {
  A(n, "appendDisc() options.center"), o = Math.max(3, Math.floor(o));
  const { normal: i, tangent: u, bitangent: p } = ye(a), d = e.vertices.length;
  e.vertices.push(R({ x: n.x, y: n.y, z: n.z }, { x: 0.5, y: 0.5 }, i, u, c));
  for (let f = 0; f <= o; ++f) {
    const l = r + f / o * s, h = Math.cos(l), m = Math.sin(l);
    e.vertices.push(
      R(
        me(n, u, p, t * h, t * m),
        { x: (h + 1) * 0.5, y: (m + 1) * 0.5 },
        i,
        u,
        c
      )
    );
  }
  for (let f = 1; f <= o; ++f)
    e.indices.push(f + d, f + d + 1, d);
}
function uo(e, n) {
  lt(e, n);
}
function lo(e, {
  center: n,
  innerRadius: t = 0.5,
  outerRadius: o = 1,
  thetaSegments: r = 32,
  phiSegments: s = 1,
  thetaStart: a = 0,
  thetaLength: c = Math.PI * 2,
  orientation: i = "xz",
  color: u
}) {
  A(n, "appendRing() options.center"), r = Math.max(3, Math.floor(r)), s = Math.max(1, Math.floor(s));
  const { normal: p, tangent: d, bitangent: f } = ye(i), l = (o - t) / s, h = [];
  for (let m = 0; m <= s; ++m) {
    const y = t + l * m, g = [];
    for (let x = 0; x <= r; ++x) {
      const _ = a + x / r * c, w = y * Math.cos(_), z = y * Math.sin(_);
      g.push(e.vertices.length), e.vertices.push(
        R(
          me(n, d, f, w, z),
          { x: (w / o + 1) * 0.5, y: (z / o + 1) * 0.5 },
          p,
          d,
          u
        )
      );
    }
    h.push(g);
  }
  for (let m = 0; m < s; ++m)
    for (let y = 0; y < r; ++y) {
      const g = h[m][y], x = h[m + 1][y], _ = h[m + 1][y + 1], w = h[m][y + 1];
      e.indices.push(g, x, w, x, _, w);
    }
}
function fo(e, {
  center: n,
  radius: t = 1,
  tube: o = 0.4,
  radialSegments: r = 12,
  tubularSegments: s = 48,
  arc: a = Math.PI * 2,
  thetaStart: c = 0,
  thetaLength: i = Math.PI * 2,
  color: u
}) {
  A(n, "appendTorus() options.center"), r = Math.max(3, Math.floor(r)), s = Math.max(3, Math.floor(s));
  const p = [];
  for (let d = 0; d <= r; ++d) {
    const f = c + d / r * i, l = Math.cos(f), h = Math.sin(f), m = [];
    for (let y = 0; y <= s; ++y) {
      const g = y / s * a, x = Math.cos(g), _ = Math.sin(g), w = t + o * l, z = {
        x: n.x + w * x,
        y: n.y + w * _,
        z: n.z + o * h
      }, E = {
        x: n.x + t * x,
        y: n.y + t * _,
        z: n.z
      }, C = k({
        x: z.x - E.x,
        y: z.y - E.y,
        z: z.z - E.z
      }), I = k({ x: -w * _, y: w * x, z: 0 });
      m.push(e.vertices.length), e.vertices.push(R(z, { x: y / s, y: d / r }, C, I, u));
    }
    p.push(m);
  }
  for (let d = 1; d <= r; ++d)
    for (let f = 1; f <= s; ++f) {
      const l = p[d][f - 1], h = p[d - 1][f - 1], m = p[d - 1][f], y = p[d][f];
      e.indices.push(l, h, y, h, m, y);
    }
}
function po(e, {
  center: n,
  radius: t = 1,
  widthSegments: o = 32,
  heightSegments: r = 16,
  phiStart: s = 0,
  phiLength: a = Math.PI * 2,
  thetaStart: c = 0,
  thetaLength: i = Math.PI,
  color: u
}) {
  A(n, "appendSphere() options.center"), o = Math.max(3, Math.floor(o)), r = Math.max(2, Math.floor(r));
  const p = Math.min(c + i, Math.PI), d = [];
  for (let f = 0; f <= r; ++f) {
    const l = [], h = f / r, m = c + h * i, y = Math.sin(m), g = Math.cos(m);
    let x = 0;
    f === 0 && c === 0 ? x = 0.5 / o : f === r && p === Math.PI && (x = -0.5 / o);
    for (let _ = 0; _ <= o; ++_) {
      const w = _ / o, z = s + w * a, E = Math.sin(z), C = Math.cos(z), I = { x: -C * y, y: g, z: E * y }, L = Math.abs(y) <= 1e-6 ? { x: 1, y: 0, z: 0 } : k({ x: E, y: 0, z: C }), q = {
        x: n.x + I.x * t,
        y: n.y + I.y * t,
        z: n.z + I.z * t
      };
      l.push(e.vertices.length), e.vertices.push(R(q, { x: w + x, y: 1 - h }, I, L, u));
    }
    d.push(l);
  }
  for (let f = 0; f < r; ++f)
    for (let l = 0; l < o; ++l) {
      const h = d[f][l + 1], m = d[f][l], y = d[f + 1][l], g = d[f + 1][l + 1];
      (f !== 0 || c > 0) && e.indices.push(h, m, g), (f !== r - 1 || p < Math.PI) && e.indices.push(m, y, g);
    }
}
function Ce(e, n, t, o, r, s, a, c, i) {
  const u = { x: 0, y: r, z: 0 }, p = { x: 1, y: 0, z: 0 }, d = [];
  for (let l = 1; l <= s; ++l)
    d.push(e.vertices.length), e.vertices.push(
      R({ x: n.x, y: n.y + o, z: n.z }, { x: 0.5, y: 0.5 }, u, p, i)
    );
  const f = e.vertices.length;
  for (let l = 0; l <= s; ++l) {
    const m = l / s * c + a, y = Math.cos(m), g = Math.sin(m);
    e.vertices.push(
      R(
        { x: n.x + t * g, y: n.y + o, z: n.z + t * y },
        { x: y * 0.5 + 0.5, y: g * 0.5 * r + 0.5 },
        u,
        p,
        i
      )
    );
  }
  for (let l = 0; l < s; ++l) {
    const h = d[l], m = f + l;
    r > 0 ? e.indices.push(m, m + 1, h) : e.indices.push(m + 1, m, h);
  }
}
function dt(e, {
  center: n,
  radiusTop: t = 1,
  radiusBottom: o = 1,
  height: r = 1,
  radialSegments: s = 32,
  heightSegments: a = 1,
  openEnded: c = !1,
  thetaStart: i = 0,
  thetaLength: u = Math.PI * 2,
  color: p
}) {
  A(n, "appendCylinder() options.center");
  const d = Number(t), f = Number(o), l = Number(r);
  s = Math.max(3, Math.floor(s)), a = Math.max(1, Math.floor(a));
  const h = l * 0.5, m = Math.abs(l) <= 1e-6 ? 0 : (f - d) / l, y = [];
  for (let g = 0; g <= a; ++g) {
    const x = [], _ = g / a, w = _ * (f - d) + d;
    for (let z = 0; z <= s; ++z) {
      const E = z / s, C = E * u + i, I = Math.sin(C), L = Math.cos(C), q = {
        x: n.x + w * I,
        y: n.y + (-_ * l + h),
        z: n.z + w * L
      }, Z = k({ x: I, y: m, z: L }), on = k({ x: L, y: 0, z: -I });
      x.push(e.vertices.length), e.vertices.push(R(q, { x: E, y: 1 - _ }, Z, on, p));
    }
    y.push(x);
  }
  for (let g = 0; g < s; ++g)
    for (let x = 0; x < a; ++x) {
      const _ = y[x][g], w = y[x + 1][g], z = y[x + 1][g + 1], E = y[x][g + 1];
      (d > 0 || x !== 0) && e.indices.push(_, w, E), (f > 0 || x !== a - 1) && e.indices.push(w, z, E);
    }
  !c && d > 0 && Ce(e, n, d, h, 1, s, i, u, p), !c && f > 0 && Ce(e, n, f, -h, -1, s, i, u, p);
}
function yo(e, {
  center: n,
  radius: t = 1,
  height: o = 1,
  radialSegments: r = 32,
  heightSegments: s = 1,
  openEnded: a = !1,
  thetaStart: c = 0,
  thetaLength: i = Math.PI * 2,
  color: u
}) {
  A(n, "appendCone() options.center"), dt(e, {
    center: n,
    radiusTop: 0,
    radiusBottom: t,
    height: o,
    radialSegments: r,
    heightSegments: s,
    openEnded: a,
    thetaStart: c,
    thetaLength: i,
    color: u
  });
}
function mo(e, {
  center: n,
  radius: t = 1,
  height: o = 1,
  capSegments: r = 4,
  radialSegments: s = 8,
  heightSegments: a = 1,
  color: c
}) {
  A(n, "appendCapsule() options.center");
  const i = Math.max(0, Number(o)), u = i * 0.5;
  r = Math.max(1, Math.floor(r)), s = Math.max(3, Math.floor(s)), a = Math.max(1, Math.floor(a));
  const p = Math.PI * t + i, d = r * 2 + a, f = s + 1;
  for (let l = 0; l <= d; ++l) {
    let h = 0, m = 0, y = 0, g = 0;
    if (l <= r) {
      const w = l / r, z = w * Math.PI / 2;
      h = -u - t * Math.cos(z), m = t * Math.sin(z), y = -t * Math.cos(z), g = w * (Math.PI * 0.5) * t;
    } else if (l <= r + a) {
      const w = (l - r) / a;
      h = -u + w * i, m = t, y = 0, g = Math.PI * 0.5 * t + w * i;
    } else {
      const w = (l - r - a) / r, z = w * Math.PI / 2;
      h = u + t * Math.sin(z), m = t * Math.cos(z), y = t * Math.sin(z), g = Math.PI * 0.5 * t + i + w * (Math.PI * 0.5) * t;
    }
    const x = p <= 1e-6 ? 0 : Math.max(0, Math.min(1, g / p));
    let _ = 0;
    l === 0 ? _ = 0.5 / s : l === d && (_ = -0.5 / s);
    for (let w = 0; w <= s; ++w) {
      const z = w / s, E = z * Math.PI * 2, C = Math.sin(E), I = Math.cos(E), L = {
        x: n.x - m * I,
        y: n.y + h,
        z: n.z + m * C
      }, q = k({
        x: -m * I,
        y,
        z: m * C
      }), Z = k({ x: m * C, y: 0, z: m * I });
      e.vertices.push(R(L, { x: z + _, y: x }, q, Z, c));
    }
    if (l > 0) {
      const w = e.vertices.length - f * 2, z = e.vertices.length - f;
      for (let E = 0; E < s; ++E) {
        const C = w + E, I = w + E + 1, L = z + E, q = z + E + 1;
        e.indices.push(C, I, L), e.indices.push(I, q, L);
      }
    }
  }
}
function ho(e, n, t, o = {}) {
  const r = e.runtime.sid(t), s = e.scenes.upsertRuntimeMaterial(n, r, {
    debugName: t,
    baseColor: o.baseColor || Y.baseColor,
    roughness: o.roughness !== void 0 ? o.roughness : Y.roughness,
    metallic: o.metallic !== void 0 ? o.metallic : Y.metallic,
    normalScale: o.normalScale !== void 0 ? o.normalScale : Y.normalScale,
    alphaCutoff: o.alphaCutoff !== void 0 ? o.alphaCutoff : Y.alphaCutoff,
    albedoTextureHash: o.albedoTextureHash || 0,
    normalTextureHash: o.normalTextureHash || 0,
    propertiesTextureHash: o.propertiesTextureHash || 0
  });
  if (!s || s.ok === !1)
    throw new Error((s == null ? void 0 : s.error) ?? `failed to upsert runtime material '${t}'`);
  return {
    hash: s.material_hash >>> 0,
    index: s.material_index >>> 0,
    rebuildRequired: s.renderer_rebuild_required === !0,
    created: s.created === !0,
    updated: s.updated === !0
  };
}
function xo(e, n, t, o) {
  const r = e.runtime.sid(t), s = e.scenes.addRuntimeMesh(n, r, o.vertexBytes, o.indices, t);
  if (!s || s.ok === !1)
    throw new Error((s == null ? void 0 : s.error) ?? `failed to register runtime mesh '${t}'`);
  return {
    hash: s.mesh_hash >>> 0,
    index: s.mesh_index >>> 0,
    rebuildRequired: s.renderer_rebuild_required === !0,
    created: s.created === !0
  };
}
function wo(e, n, t, o) {
  const { ECS: r, Components: s } = tn(), { Transform: a, MeshRenderer: c, Name: i } = s;
  typeof r.bindModule == "function" && r.bindModule(e);
  const u = r.createEntity(t);
  o.name && (i == null ? void 0 : i.typeId) !== void 0 && r.writeComponent(t, u, i, {
    name_hash: n.runtime.sid(o.name)
  });
  const p = r.writeComponent(t, u, a, {
    position: o.position || { x: 0, y: 0, z: 0 },
    rotation: o.rotation || { x: 0, y: 0, z: 0, w: 1 },
    scale: o.scale || { x: 1, y: 1, z: 1 }
  }), d = r.writeComponent(t, u, c, {
    mesh: o.meshHash >>> 0,
    material: o.materialHash >>> 0,
    color: o.color || { x: 255, y: 255, z: 255, w: 255 }
  });
  return {
    entityId: u >>> 0,
    transformPtr: p.ptr >>> 0,
    meshRendererPtr: d.ptr >>> 0
  };
}
function zo(e, n) {
  if (!e) return;
  const t = at(), { Components: o } = tn();
  o.Transform.write(new DataView(t.HEAPU8.buffer), e >>> 0, n);
}
export {
  nn as RUNTIME_VERTEX_STRIDE_BYTES,
  ao as appendBox,
  mo as appendCapsule,
  uo as appendCircle,
  yo as appendCone,
  dt as appendCylinder,
  lt as appendDisc,
  co as appendPlane,
  io as appendPyramid,
  H as appendQuad,
  lo as appendRing,
  po as appendSphere,
  fo as appendTorus,
  Q as appendTriangle,
  kn as applyPreviewEvent,
  ke as bindEngine,
  $n as bootMiniEngineRuntime,
  fe as clamp,
  Mt as clearGameplayComponents,
  wt as createGameplayComponentLibrary,
  so as createMeshBuilder,
  Ct as createScene,
  xt as defineGameplayComponent,
  Ye as degToRad,
  It as destroyScene,
  Tt as disableGameAudio,
  Wt as disposeAudio,
  Pt as enableGameAudio,
  ho as ensureRuntimeMaterial,
  ro as finalizeMesh,
  Kt as forwardFromYaw,
  Dn as frameMiniEngine,
  Fe as gameplayComponentsNeedRuntimeSync,
  _n as getAssetScenes,
  Bt as getAudioContext,
  yt as getCameraEntityId,
  Te as getEcsBindings,
  Mn as getEngineAssets,
  ft as getEngineBaseUrl,
  gt as getGameplayComponentSchemaById,
  zt as getGameplayComponentSchemaByName,
  un as getGameplayComponentSchemas,
  Dt as getLoadedSounds,
  jt as getMasterVolume,
  v as getMini,
  Ot as getPlayingCount,
  G as getSceneHandle,
  rn as getSceneSaveHook,
  Ht as getSoundInfo,
  Pe as getWasmModule,
  $e as hideSourceAssetEntities,
  Oe as initAudio,
  Ge as instantiateAsset,
  pt as isEngineReady,
  Vt as isPlaying,
  qt as isSoundLoaded,
  Yt as length2D,
  F as lerp,
  En as loadAssetFromBuffer,
  J as loadAssetFromUrl,
  Wn as loadAudioFromBuffer,
  On as loadAudioFromUrl,
  je as loadProjectAssetsIntoEngine,
  Nt as loadProjectAudioAssets,
  In as loadScene,
  vt as loadSceneFromUrl,
  At as loadSceneFull,
  Zt as math,
  Et as multiplayer,
  to as normalizeQuat,
  oo as packColor,
  Ee as packSignedRgb10a2Unorm,
  Ae as packUvUnorm16,
  St as pauseSound,
  bt as playMusic,
  Yn as playSound,
  Jt as quat,
  et as quatClone,
  Xe as quatCreate,
  tt as quatFromAxisAngle,
  ot as quatFromYawPitch,
  Je as quatIdentity,
  en as quatLookRotation,
  nt as quatMultiply,
  S as quatNormalize,
  Ze as quatRotateVector,
  st as quatSlerp,
  Ke as radToDeg,
  _t as registerGameplayComponents,
  xo as registerRuntimeMesh,
  Hn as resizeMiniEngineRenderer,
  An as restoreSourceAssetEntities,
  Lt as resumeAudio,
  Ut as resumeSound,
  Qt as rightFromYaw,
  Rn as saveScene,
  vn as saveSceneFull,
  eo as scalar,
  mt as setCameraEntityId,
  Rt as setMainScene,
  Gt as setMasterVolume,
  Le as setSceneHandle,
  ht as setSceneSaveHook,
  $t as setSoundVolume,
  wo as spawnRenderable,
  kt as startGame,
  Kn as stopAll,
  Ft as stopMusic,
  de as stopSound,
  be as syncGameplayComponentsWithRuntime,
  no as transform,
  zo as updateTransform,
  Xt as vec3,
  oe as vec3Add,
  Qe as vec3Clone,
  Qn as vec3Create,
  D as vec3Cross,
  Zn as vec3Distance,
  Xn as vec3Dot,
  pe as vec3Length,
  Jn as vec3Lerp,
  k as vec3Normalize,
  se as vec3Scale,
  j as vec3Subtract,
  rt as vec3TransformByQuaternion,
  it as writeVertex
};
